### ISSUE TITLE
glide up to a new version that removes a package

### ISSUE THREAD

**Issue by arschles** [Masterminds/glide#782]
If I start with this:

- My code depends on version 1 of `github.com/my/package`
- I `import github.com/my/package/one/two`

Then I `glide up github.com/my/package` to a new version that doesn't contain `github.com/my/package/one/two`, the command will fail because my code references a directory that doesn't exist. We saw a similar circumstance in a [kubernetes-incubator project](https://github.com/kubernetes-incubator/service-catalog) recently, when we were trying to upgrade from an old version of `k8s.io/client-go` to a more modern version.

Our workaround was to do the following:

1. Copy the existing `glide.yaml` and `glide.lock` to a new, empty directory
2. `glide up`
3. Copy the new `glide.lock` and `vendor` directory back to the original directory (overwriting the existing `glide.lock` and `vendor`)

After we did this, the `go build` would fail as expected and we could go on with our work to complete the upgrade.

Is there any way to tell `glide up` to not walk the dependency tree of my code, instead only looking at the dependency tree expressed in my `glide.yaml` file? We are fine with the workaround listed above, but I'm curious if there's a better way.

**Comment by technosophos** [comment]
When you did the workaround, did it delete a bunch of dependencies? I was looking into the possibility of doing this, and the main issue is that if it doesn't scan your current codebase, any transitive dependencies that are not in glide.yaml will ignored.

**Comment by arschles** [comment]
@technosophos Sorry, I don't have something concrete here. I didn't explicitly look, but I did copy over the `glide.lock` before running, and it didn't change much except for version numbers as I expected. If necessary, I can reproduce the whole thing -- just let me know.

I had a few other ideas in my head, so I'll put there here:

- Can I pass glide a flag to only check the top-level package imports in my code?
    - For example, if my code has an import like this: `import "github.com/my/package/one/two"`, can a command like `glide up --top-level-only` ensure only that `github.com/my/package` is present?
    - It could be problematic to determine when a package name is fully-qualified and references the top level of the package, unless we already have it on disk
- Can I tell glide to walk the dependency trees and fetch transitive dependencies for only the code that lives in my `vendor/`?
    - For example, `glide up --vendor-only`
    - This would allow me to do version upgrades first, then I can `glide get` other packages that my code needs
- Can I tell glide to upgrade only the things in my `glide.lock`?
    - For example, `glide up --lock-only`
    - Just as with `glide up --vendor-only`, this would effectively allow me to update what I already have, and then `glide get` what's missing later
    - This could be also be useful for rebuilding a lost `glide lock`

Anyway, that's way more than $0.02. Hopefully it helps a bit.

**Comment by mattfarina** [comment]
@arschles there's a problem of grabbing the right transitive dependencies. For example, Glide only grabs the transitive dependencies that are in use. But, if it can't walk the import tree (only grabbing top level) then it doesn't know which to import. This is dangers with Kubernetes related things because importing all possibilities can accidentally import all possible Kubernetes ones... which can take a very long time.

Any ideas on how to make the experience better given that?

**Comment by arschles** [comment]
@mattfarina good point, one possibility is to have glide warn, but not fail, when it finds `import`s in my code that

1. Has a dependent package in the package name (i.e. `import github.com/my/package/a/b/c` and `github.com/my/package` is in the `glide.yaml`)
2. `import`s a package that doesn't exist

If it did these warnings, it seems like `glide` could avoid pulling in a massive set of dependencies, but still tell the user that their build is likely to fail.

I may be missing something here. I'm looking forward to hearing your thoughts.
